File: test/scripts/23-thread-pools/correct.py
#!/usr/bin/env python3
"""
CORRECT: ThreadPoolExecutor - reuse threads.
Resume: "Use Thread Pools"
"""
import os
import signal
import time
from concurrent.futures import ThreadPoolExecutor
TASKS = 500
DURATION_SEC = int(os.environ.get("DURATION_SEC", "60"))
stop = False

def task(_):
    return sum(range(1000))

def main():
    global stop
    print("CORRECT: ThreadPoolExecutor")
    signal.signal(signal.SIGINT, lambda s, f: globals().__setitem__('stop', True))
    start = time.perf_counter()
    laps = 0
    with ThreadPoolExecutor(max_workers=8) as ex:
        while not stop and (time.perf_counter() - start) < DURATION_SEC:
            list(ex.map(task, range(TASKS)))
            laps += 1
    print(f"Stopped: {laps} laps")

if __name__ == "__main__":
    main()

File: test/scripts/23-thread-pools/correct.js
const { Worker } = require('worker_threads');
const NUM_WORKERS = 8;  // Pool size = cores
const TASKS_PER_WORKER = 13;  // 8*13 ~ 100 tasks
const DURATION_SEC = parseInt(process.env.DURATION_SEC || '60', 10);
let laps = 0;
const start = Date.now();
const code = `let t=0;for(let i=0;i<${1000*TASKS_PER_WORKER};i++)t+=i;`;
console.log('CORRECT: 8 workers (pool) vs 100');
function run() {
  if (Date.now() - start >= DURATION_SEC * 1000) {
    console.log('Stopped:', laps, 'laps');
    return;
  }
  const workers = [];
  for (let i = 0; i < NUM_WORKERS; i++) {
    workers.push(new Promise((r) => {
      const w = new Worker(code, { eval: true });
      w.on('exit', r);
    }));
  }
  Promise.all(workers).then(() => { laps++; setImmediate(run); });
}
run();
process.on('SIGINT', () => process.exit(0));
